zboxcore/sdk/dirworker.go

```diff
@@ -4,66 +4,123 @@ import (
 	"bytes"
 	"context"
 	"errors"
+	"fmt"
 	"io/ioutil"
+	"math/bits"
 	"mime/multipart"
 	"net/http"
-	"path/filepath"
 	"strings"
+	"sync"
 	"time"
 
+	"github.com/0chain/gosdk/zboxcore/allocationchange"
 	"github.com/0chain/gosdk/zboxcore/blockchain"
+	"github.com/0chain/gosdk/zboxcore/client"
 	l "github.com/0chain/gosdk/zboxcore/logger"
 	"github.com/0chain/gosdk/zboxcore/zboxutil"
 )
 
+const (
+	DirectoryExists = "directory_exists"
+)
+
 type DirRequest struct {
 	allocationID string
-	name         string
+	allocationTx string
+	remotePath   string
+	blobbers     []*blockchain.StorageNode
 	ctx          context.Context
-	action       string // create, del
+	wg           *sync.WaitGroup
+	dirMask      uint32
+	mu           *sync.Mutex
 	connectionID string
 	Consensus
 }
 
 func (req *DirRequest) ProcessDir(a *Allocation) error {
 	numList := len(a.Blobbers)
+	req.wg = &sync.WaitGroup{}
+	req.wg.Add(numList)
 
 	l.Logger.Info("Start creating dir for blobbers")
 
-	await := make(chan error, numList)
-
 	for i := 0; i < numList; i++ {
 		go func(blobberIdx int) {
-			err := req.createDirInBlobber(a.Blobbers[blobberIdx])
+			defer req.wg.Done()
+
+			err := req.createDirInBlobber(a.Blobbers[blobberIdx], blobberIdx)
 			if err != nil {
 				l.Logger.Error(err.Error())
+				return
 			}
-			await <- err
 		}(i)
 	}
 
-	msgList := make([]string, 0, numList)
-	for i := 0; i < numList; i++ {
-		err := <-await
-		if err != nil {
-			msgList = append(msgList, err.Error())
-		}
+	req.wg.Wait()
+
+	if !req.isConsensusOk() {
+		return errors.New("Directory creation failed due to consensus not met")
 	}
 
-	if len(msgList) > 0 {
-		return errors.New(strings.Join(msgList, ", "))
+	writeMarkerMU, err := CreateWriteMarkerMutex(client.GetClient(), a)
+	if err != nil {
+		return fmt.Errorf("Directory creation failed. Err: %s", err.Error())
+	}
+	err = writeMarkerMU.Lock(context.TODO(), req.connectionID)
+	defer writeMarkerMU.Unlock(context.TODO(), req.connectionID) //nolint: errcheck
+	if err != nil {
+		return fmt.Errorf("Directory creation failed. Err: %s", err.Error())
 	}
 
+	req.consensus = 0
+	wg := &sync.WaitGroup{}
+	okBlobbers := bits.OnesCount32(req.dirMask)
+	wg.Add(okBlobbers)
+	commitReqs := make([]*CommitRequest, okBlobbers)
+	var c, pos int
+	for i := req.dirMask; i != 0; i &= ^(1 << pos) {
```

Comment:
Please wraps the dirMask binary actions into a separate method and name it properly to make it easier to read. Since the main purpose is to get all the blobbers that are in active set in dirMask, we can have a method named: `getActiveBlobbers() []*blobberNodes`, and put all the binary action inside the method. 


Also, please move the whole commit requests code into a method, the DirRequest is already long enough.